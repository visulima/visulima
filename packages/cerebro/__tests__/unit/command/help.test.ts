import type { Mock } from "vitest";
import { beforeEach, describe, expect, it, vi } from "vitest";

import type { Command as ICommand, Toolbox as IToolbox } from "../../../src/@types";
import HelpCommand from "../../../src/command/help";
import globalOptions from "../../../src/default-options";
import commandLineUsage from "../../../src/util/command-line-usage";

vi.mock("../../../src/util/command-line-usage");
vi.mock("../default-options", () => {
    return {
        default: [
            // Your default options structure here
        ],
    };
});

const loggerMock = {
    debug: vi.fn(),
    raw: vi.fn(),
    warn: vi.fn(),
};

const runtimeMock = {
    getCliName: vi.fn().mockReturnValue("testcli"),
    getCommandSection: vi.fn().mockReturnValue({ footer: "Test footer", header: "Test header" }),
};

const headerSection = {
    content: "{cyan testcli} {green <command>} [positional arguments] {yellow [options]}",
    header: "{inverse.cyan  Usage }",
};

const footerSection = {
    content: 'Run "{cyan testcli} {green help <command>}" or "{cyan testcli} {green <command>} {yellow --help}" for help with a specific command.',
    raw: true,
};

const globalOptionsOptionsList = {
    header: "{inverse.yellow  Global Options }",
    optionList: globalOptions,
};

describe("command/help", () => {
    const commandsMap = new Map<string, ICommand>([]);

    beforeEach(() => {
        vi.clearAllMocks();

        commandsMap.clear();

        commandsMap.set("test", {
            description: "A test command",
            execute: () => {},
            group: "test",
            name: "test",
            options: [{ description: "Enable verbose output", name: "verbose" }],
        });
    });

    it("should print general help", () => {
        expect.assertions(3);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenCalledWith("Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenCalledWith("Test footer");
    });

    it("should print command-specific help", () => {
        expect.assertions(3);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "test",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenCalledWith("Test header");
        expect(commandLineUsage).toHaveBeenCalledWith([
            {
                content: "{cyan testcli} {green test} [options]",
                header: "{inverse.cyan  Usage }",
            },
            {
                content: "A test command",
                header: "{inverse.green  Description }",
            },
            {
                header: "{inverse.yellow  Command Options }",
                optionList: [
                    {
                        description: "Enable verbose output",
                        name: "verbose",
                    },
                ],
            },
            globalOptionsOptionsList,
        ]);
        expect(loggerMock.raw).toHaveBeenCalledWith("Test footer");
    });

    it("should print general help and not include hidden commands", () => {
        expect.assertions(4);

        const secretCommandsMap = new Map([
            // Add a hidden command for testing
            [
                "secret",
                {
                    execute: () => {},
                    hidden: true,
                    name: "secret",
                },
            ],
            [
                "test",
                {
                    description: "A test command",
                    execute: () => {},
                    name: "test",
                    options: [{ description: "Enable verbose output", name: "verbose" }],
                },
            ],
        ]);

        const helpCommand = new HelpCommand(secretCommandsMap);
        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenCalledWith("Test header");
        expect(commandLineUsage).toHaveBeenCalledWith([
            headerSection,
            {
                content: [["{green test} ", "A test command"]],
                header: "{inverse.green  Available Commands }",
            },
            globalOptionsOptionsList,
            footerSection,
        ]);
        expect(loggerMock.raw).toHaveBeenCalledWith("Test footer");

        // Ensure hidden commands are not included in the output
        const usageCalls = (commandLineUsage as Mock).mock.calls[0][0];

        expect(usageCalls.some((section) => section.content?.includes("secret"))).toBeFalsy();
    });

    it("should display aliases if present", () => {
        expect.assertions(3);

        const commandWithAlias = {
            alias: ["cwa", "aliasForCommand"],
            execute: () => {},
            name: "commandWithAlias",
        };
        commandsMap.set(commandWithAlias.name, commandWithAlias);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "commandWithAlias",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenCalledWith("Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenCalledWith("Test footer");
    });

    it("should display command examples", () => {
        expect.assertions(3);

        const commandWithExamples = {
            examples: ["testcli test --verbose", "testcli test --verbose --debug"],
            execute: () => {},
            name: "commandWithExamples",
        };

        commandsMap.set(commandWithExamples.name, commandWithExamples);

        const helpCommand = new HelpCommand(commandsMap);
        const toolboxMock = {
            commandName: "commandWithExamples",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenCalledWith("Test header");
        expect(commandLineUsage).toHaveBeenCalledWith([
            {
                content: "{cyan testcli} {green commandWithExamples}",
                header: "{inverse.cyan  Usage }",
            },
            globalOptionsOptionsList,
            {
                content: ["testcli test --verbose", "testcli test --verbose --debug"],
                header: "Examples",
            },
        ]);
        expect(loggerMock.raw).toHaveBeenCalledWith("Test footer");
    });

    it("should display a empty help command if no commands exists", () => {
        expect.assertions(3);

        const helpCommand = new HelpCommand(new Map());

        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenCalledWith("Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenCalledWith("Test footer");
    });

    it("should display group of commands if the group option is passed", () => {
        expect.assertions(3);

        commandsMap.set("test2", {
            description: "A test2 command",
            execute: () => {},
            group: "test",
            name: "test2",
            options: [{ description: "Enable verbose output", name: "verbose" }],
        });

        commandsMap.set("no-group-test", {
            description: "A test command",
            execute: () => {},
            name: "no-group-test",
            options: [{ description: "Enable verbose output", name: "verbose" }],
        });

        const helpCommand = new HelpCommand(commandsMap);

        const toolboxMock = {
            commandName: "help",
            logger: loggerMock,
            options: {
                group: "test",
            },
            runtime: runtimeMock,
        };

        helpCommand.execute(toolboxMock as unknown as IToolbox);

        expect(loggerMock.raw).toHaveBeenCalledWith("Test header");
        expect(commandLineUsage).toMatchSnapshot();
        expect(loggerMock.raw).toHaveBeenCalledWith("Test footer");
    });
});
