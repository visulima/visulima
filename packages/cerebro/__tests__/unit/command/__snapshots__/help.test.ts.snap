// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`command/help > should display a empty help command if no commands exists 1`] = `
[MockFunction commandLineUsage] {
  "calls": [
    [
      [
        {
          "content": "{cyan testcli} {green <command>} [positional arguments] {yellow [options]}",
          "header": "{inverse.cyan  Usage }",
        },
        {
          "header": "{inverse.yellow  Global Options }",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": "Run "{cyan testcli} {green help <command>}" or "{cyan testcli} {green <command>} {yellow --help}" for help with a specific command.",
          "raw": true,
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should display aliases if present 1`] = `
[MockFunction commandLineUsage] {
  "calls": [
    [
      [
        {
          "content": "{cyan testcli} {green commandWithAlias}",
          "header": "{inverse.cyan  Usage }",
        },
        {
          "content": [
            "cwa",
            "aliasForCommand",
          ],
          "header": "Alias(es)",
        },
        {
          "header": "{inverse.yellow  Global Options }",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should display group of commands if the group option is passed 1`] = `
[MockFunction commandLineUsage] {
  "calls": [
    [
      [
        {
          "content": "{cyan testcli} {green <command>} [positional arguments] {yellow [options]}",
          "header": "{inverse.cyan  Usage }",
        },
        {
          "content": [
            [
              "{green test} ",
              "A test command",
            ],
            [
              "{green test2} ",
              "A test2 command",
            ],
          ],
          "header": "{inverse.green  Available Test Commands }",
        },
        {
          "header": "{inverse.yellow  Global Options }",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": "Run "{cyan testcli} {green help <command>}" or "{cyan testcli} {green <command>} {yellow --help}" for help with a specific command.",
          "raw": true,
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`command/help > should print general help 1`] = `
[MockFunction commandLineUsage] {
  "calls": [
    [
      [
        {
          "content": "{cyan testcli} {green <command>} [positional arguments] {yellow [options]}",
          "header": "{inverse.cyan  Usage }",
        },
        {
          "content": [
            [
              "{green test} ",
              "A test command",
            ],
          ],
          "header": " {inverse.green  Test }",
        },
        {
          "header": "{inverse.yellow  Global Options }",
          "optionList": [
            {
              "description": "Turn on verbose output",
              "group": "global",
              "name": "verbose",
              "type": [Function],
            },
            {
              "description": "Turn on debugging output",
              "group": "global",
              "name": "debug",
              "type": [Function],
            },
            {
              "alias": "h",
              "description": "Print out helpful usage information",
              "group": "global",
              "name": "help",
              "type": [Function],
            },
            {
              "alias": "q",
              "description": "Silence output",
              "group": "global",
              "name": "quiet",
              "type": [Function],
            },
            {
              "alias": "V",
              "description": "Print version info",
              "group": "global",
              "name": "version",
              "type": [Function],
            },
            {
              "description": "Turn off colored output",
              "group": "global",
              "name": "no-color",
              "type": [Function],
            },
            {
              "description": "Force colored output",
              "group": "global",
              "name": "color",
              "type": [Function],
            },
          ],
        },
        {
          "content": "Run "{cyan testcli} {green help <command>}" or "{cyan testcli} {green <command>} {yellow --help}" for help with a specific command.",
          "raw": true,
        },
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
